types/i18next/index.d.ts

```diff
@@ -1,157 +1,601 @@
-        parseMissingKeyHandler?: (key: string) => void;
-        appendNamespaceToMissingKey?: boolean;
-        postProcess?: string | any[];
-        returnNull?: boolean;
-        returnEmptyString?: boolean;
-        returnObjects?: boolean;
-        returnedObjectHandler?: (key: string, value: string, options: any) => void;
-        joinArrays?: string;
-        overloadTranslationOptionHandler?: (args: any[]) => TranslationOptions;
-        interpolation?: InterpolationOptions;
-        detection?: any;
-        backend?: any;
-        cache?: any;
+    type Callback = (error: any, t: TranslationFunction) => void;
+
+    type TranslationFunction<TResult = string, TValues extends object = object, TKeys extends string = string> = (key: TKeys | TKeys[], options?: TranslationOptions<TValues>) => TResult;
+
+    interface Resource {
+        [language: string]: ResourceLanguage;
     }
 
-    // init options for react-i18next
-    interface ReactOptions {
-      wait?: boolean;
+    interface ResourceLanguage {
+        [namespace: string]: ResourceKey;
     }
 
-    type TranslationFunction = (key: string, options?: TranslationOptions) => string;
-    type LoadCallback = (error: any, t: TranslationFunction) => void;
-
-    interface I18n {
-        // api
-        init(options?: Options & ReactOptions, callback?: LoadCallback): I18n;
-        use(module: any): I18n;
-        t(key: string, options?: TranslationOptions): string | any | any[];
-        exists(key: string, options?: TranslationOptions): boolean;
-        getFixedT(lng?: string, ns?: string | string[]): TranslationFunction;
-        changeLanguage(lng: string, callback?: LoadCallback): void;
+    interface ResourceKey {
+        [key: string]: any;
+    }
+
+    interface i18n {
+        /**
+         * The default export of the i18next module is an i18next instance ready to be initialized by calling init.
+         * You can create additional instances using the createInstance function.
+         *
+         * @param options - Initial options.
+         * @param callback - will be called after all translations were loaded or with an error when failed (in case of using a backend).
+         */
+        init(options: InitOptions, callback?: Callback): i18n;
+        init(callback?: Callback): i18n;
+
+        loadResources(callback?: (err: any) => void): void;
+
+        /**
+         * The use function is there to load additional plugins to i18next.
+         * For available module see the plugins page and don't forget to read the documentation of the plugin.
+         */
+        use(module: any): i18n;
+
+        /**
+         * Please have a look at the translation functions like interpolation, formatting and plurals for more details on using it.
+         */
+        t<TResult = string, TValues extends object = object, TKeys extends string = string>(key: TKeys | TKeys[], options?: TranslationOptions<TValues>): TResult;
```

Comment:
So, `TranslationFunction` could be changed to:

```
interface TranslationFunction<T> {
    <TResult extends T = T, TValues extends object = object, TKeys extends string = string>(key: TKeys | TKeys[], options?: TValues): TResult
}
```

All translation function occurrences should use `TranslationFunction<any>` type like:
```
t: TranslationFunction<any>;
```
and
```
getFixedT(lng: string | string[], ns?: string | string[]): TranslationFunction<any>;
```

And exists function should be defined as:

```
exists: TranslationFunction<boolean>;
```